MSYS2 build failures

### Experienced behavior

Current master does not build on MSYS2 as documented.

```
$ ninja
[1/67] Compiling c object src/graphene-1.0@sha/graphene-alloc.c.obj
FAILED: src/graphene-1.0@sha/graphene-alloc.c.obj
cc @src/graphene-1.0@sha/graphene-alloc.c.obj.rsp
../src/graphene-alloc.c: In function 'graphene_aligned_alloc':
../src/graphene-alloc.c:94:11: error: implicit declaration of function 'posix_memalign' [-Werror=implicit-function-declaration]
   errno = posix_memalign (&res, alignment, real_size);
           ^~~~~~~~~~~~~~
cc1.exe: some warnings being treated as errors
[2/67] Compiling c object src/graphene-1.0@sha/graphene-euler.c.obj
[3/67] Compiling c object src/graphene-1.0@sha/graphene-box.c.obj
../src/graphene-box.c:53:0: warning: "_WIN32_WINNT" redefined
 #define _WIN32_WINNT 0x0600

In file included from C:/msys64/mingw64/x86_64-w64-mingw32/include/crtdefs.h:10:0,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/stdlib.h:9,
                 from ../src/graphene-private.h:30,
                 from ../src/graphene-box.c:33:
C:/msys64/mingw64/x86_64-w64-mingw32/include/_mingw.h:225:0: note: this is the location of the previous definition
 #define _WIN32_WINNT 0x502

../src/graphene-box.c:688:15: warning: no previous prototype for 'InitBoxFunc' [-Wmissing-prototypes]
 BOOL CALLBACK InitBoxFunc (PINIT_ONCE InitOnce,
               ^~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

### Expected behavior

Ninja completes the build so I can make you a nice mingw-w64-graphene-git package to allow the new GTK3 dependencies to be built on Windows.

### Steps to reproduce

These also form updated build instructions for MSYS2.

From an MSYS2 environment:

1. <s>Build [mingw-w64-meson](https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-meson) with `makepkg-mingw -sf` (packages are not yet available in the repository because @alexpux has been busy)</s> Now just install them: `pacman -S mingw-w64-i686-meson mingw-w64-x86_64-meson`
2. <s>Install the resultant packages: `pacman -U *.pkg.tar.xz`</s>
3. <s>Install ninja from the repository: `pacman -S mingw-w64-{x86_64,i686}-ninja`</s> Now pulled in as a dependency.
4. Install graphene's documented dependencies:

    ```
    pacman -Syyuu    # repeat as needed, closing the terminal when it tells you

    pacman -S automake autoconf libtool pkg-config gtk-doc
    pacman -S glib2   # why not native (mingw-w64-{x86_64,i686}) ?
    ```

5. From a MINGW64 or a MINGW32 shell,

    ```
    cd /usr/src/
    git clone git://github.com/ebassi/graphene
    cd graphene/
    mkdir _build && cd _build
    meson.py ..     # note the different name on these platforms
    ninja
    ```

At which point it fails with the error above.

### Operating system in use

Windows 7 Professional N (on Virtualbox on Debian)

### SIMD implementation in use

Not a clue, I only care about getting graphene built sensibly so I can fix wacom stylus tilt in GTK master.

--- a/meson.build
+++ b/meson.build
@@ -129,10 +129,21 @@ conf.set('HAVE_STDBOOL_H', cc.has_header('stdbool.h'))
 conf.set('HAVE_MEMORY_H', cc.has_header('memory.h'))
 
 # Functions
-conf.set('HAVE_ALIGNED_ALLOC', cc.has_function('aligned_alloc', prefix: '#include <stdlib.h>'))
-conf.set('HAVE__ALIGNED_MALLOC', cc.has_function('_aligned_malloc', prefix: '#include <malloc.h>'))
-conf.set('HAVE_MEMALIGN', cc.has_function('memalign', prefix: '#include <stdlib.h>\n#include <malloc.h>'))
-conf.set('HAVE_POSIX_MEMALIGN', cc.has_function('posix_memalign', prefix: '#include <stdlib.h>'))
+if cc.has_function('memalign', prefix: '#include <stdlib.h>\n#include <malloc.h>')
+  conf.set('HAVE_MEMALIGN', 1)
+elif cc.has_function('_aligned_malloc', prefix: '#include <malloc.h>')
+  conf.set('HAVE__ALIGNED_MALLOC', 1)
+# Don't probe the ones below on Windows because when building with
+# MinGW-w64 on MSYS2, Meson<0.37.0 incorrectly detects those below as
+# being available even though they're not.
+elif cc.has_function('aligned_alloc', prefix: '#include <stdlib.h>') and not (host_system == 'windows')
+  conf.set('HAVE_ALIGNED_ALLOC', 1)
+elif cc.has_function('posix_memalign', prefix: '#include <stdlib.h>') and not (host_system == 'windows')
+  conf.set('HAVE_POSIX_MEMALIGN', 1)
+else
+  error('No aligned malloc function could be found.')
+endif
+
 conf.set('HAVE_SINCOSF', cc.has_function('sincosf', prefix: '#define _GNU_SOURCE\n#include <math.h>'))
 
 # Debugging

